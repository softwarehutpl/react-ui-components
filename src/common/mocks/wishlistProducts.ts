export default [
  {
    name: 'Blue summer hat',
    price: '25.00 EUR',
    image: 'https://cdn.pixabay.com/photo/2015/10/13/03/39/fashion-985556_1280.jpg',
  },
  {
    name: 'Huge sombrero',
    price: '30.99 EUR',
    image: 'https://cdn.pixabay.com/photo/2015/12/08/01/04/sombrero-1082322_1280.jpg',
  },
  {
    name: 'Hunter hat',
    price: '23.50 EUR',
    image: 'https://cdn.pixabay.com/photo/2016/11/18/14/15/forest-1834831_1280.jpg',
  }
]
