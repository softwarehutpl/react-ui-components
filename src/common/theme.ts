const theme: { [key: string]: { [key: string]: { [key: string]: string } } } = {
  colors: {
    primary: {
      base: '#000000',
      light: '#ffffff',
      dark: '#a8b3bd',
    },
    secondary: {
      base: '#3690e3',
      light: '#75b0e6',
      dark: '#034d91',
    },
    error: {
      base: '#f44336',
      light: '#f56c62',
      dark: '#a1140a',
    },
    warning: {
      base: '#faa72d',
      light: '#ffc36b',
      dark: '#995b00',
    },
    info: {
      base: '#7979d4',
      light: '#9797d1',
      dark: '#4c4cc7',
    },
    success: {
      base: '#4caf50',
      light: '#88c28a',
      dark: '#124f14',
    },
    disabled: {
      base: '#d9d9d9',
      light: '#f0eded',
      dark: '#8c8989',
    },
  },
  // TODO: change it in dedicated task for that (change ease-out for example and add more options)
  transitions: {
    all: {
      fast: 'all 0.5s ease-out',
      mid: 'all 2s ease-out',
      slow: 'all 4s ease-out',
    },
    background: {
      fast: 'background-color 0.5s ease-out',
      mid: 'background-color 2s ease-out',
      slow: 'background-color 4s ease-out',
    },
    fontColor: {
      fast: 'color 0.5s ease-out',
      mid: 'color 2s ease-out',
      slow: 'color 4s ease-out',
    }
  }
};

export default theme;
