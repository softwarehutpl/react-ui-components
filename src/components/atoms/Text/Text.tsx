import React from 'react';
import styled from 'styled-components';
import { COLOR_OPTIONS, TEXT_ALIGNMENT_OPTIONS, FONT_WEIGHT_OPTIONS } from '../../../common/constants/consts';

interface IText {
  className?: string;
  children: string;
  color?: string;
  fontColor?: string;
  fontWeight?:
    | 'normal'
    | 'bold'
    | '100'
    | '200'
    | '300'
    | '400'
    | '500'
    | '600'
    | '700'
    | '800'
    | '900';
  lineHeight?: number;
  fontSize?: number;
  textAlign?: 'left' | 'right' | 'center' | 'justify';
}

const Text = ({ className, children }: IText) => <p className={className}>{children}</p>;

const defaultProps = {
  color: COLOR_OPTIONS.primary,
  fontSize: 16,
  fontWeight: FONT_WEIGHT_OPTIONS.normal,
  lineHeight: 1.5,
  textAlign: TEXT_ALIGNMENT_OPTIONS.justify,
};

Text.defaultProps = defaultProps;

export default styled(Text)<IText>`
  ${({
    theme,
    color = defaultProps.color,
    fontColor,
    fontSize = defaultProps.fontSize,
    fontWeight = defaultProps.fontWeight,
    lineHeight = defaultProps.lineHeight,
    textAlign = defaultProps.textAlign,
  }) => ({
    color: fontColor || theme.colors[color].base,
    'font-size': `${fontSize}px`,
    'line-height': `${lineHeight}`,
    'font-weight': fontWeight,
    'text-align': textAlign,
  })}
`;
