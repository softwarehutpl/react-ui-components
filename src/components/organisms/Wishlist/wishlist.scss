.wishlistItem {
  height: 120px;
  display: flex;
  padding: 10px;
  &:hover {
    box-shadow: 0px 4px 10px 4px rgba(217,217,217,0.6);
  }
}

.wishlistItem__imageWrapper {
  height: 100%;
  width: 120px;
  margin-right: 10px;
}

.wishlistItem__textWrapper {
  display: flex;
  flex-direction: column;
  justify-content: center;
  width: 100%;
  position: relative;
}

.wishlistItem__cartIcon {
  position: absolute;
  bottom: 0;
  right: 0;
}
