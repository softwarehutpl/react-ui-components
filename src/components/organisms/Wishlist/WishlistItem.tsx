import React from 'react';
import styled from 'styled-components';
import Text, {FontWeights, IText} from '../../atoms/Text/Text';
import Image from '../../atoms/Image/Image';
import './wishlist.scss';
import { FONT_WEIGHT_OPTIONS } from '../../../common/constants/consts';
import { DeleteForever } from '@styled-icons/material-sharp/DeleteForever';

export interface IWishlistProduct {
  name: string;
  price: string;
  image: string;
}

export interface IWishlistCommon {
  itemClassName?: string;
  hideHeartDislike?: boolean;
  hideCartIcon?: boolean;
  color?: string;
  onClick?: () => void;
  nameTextProps?: IText;
  priceTextProps?: IText;
  onDeleteIconClick?: () => void;
  onItemClick?: () => void;
  deleteIconColor?: string;
  hideDeleteIcon?: boolean;
}

interface IWishlistItem extends IWishlistCommon {
  product: IWishlistProduct;
}

export const StyledWishlistDeleteIcon = styled(DeleteForever)<{ color?: string, deleteIconColor?: string }>`${({
  theme, color = defaultProps.color, deleteIconColor
}) => ({
  color: deleteIconColor || theme.colors[color].base,
  width: '30px',
  position: 'absolute',
  bottom: '0',
  right: '0',
  cursor: 'pointer',
})}`

const WishlistItem = ({ product, color, nameTextProps, priceTextProps, itemClassName, onDeleteIconClick, onItemClick, deleteIconColor, hideDeleteIcon }: IWishlistItem) => {
  const nameProps = {
    color,
    fontWeight: FONT_WEIGHT_OPTIONS['600'] as FontWeights,
    ...nameTextProps,
  };
  const priceProps = {
    color,
    ...priceTextProps,
  };
  return (
    <div className={`${itemClassName} wishlistItem`} onClick={onItemClick}>
      <div className="wishlistItem__imageWrapper">
        <Image src={product.image} alt={product.name} />
      </div>
      <div className="wishlistItem__textWrapper">
        <Text {...nameProps}>{product.name}</Text>
        <Text {...priceProps}>{product.price}</Text>
        { !hideDeleteIcon && (
          <StyledWishlistDeleteIcon color={color} deleteIconColor={deleteIconColor} onClick={(e: React.MouseEvent) => {
            e.stopPropagation();
            onDeleteIconClick && onDeleteIconClick();
          }} />
        )}
      </div>
    </div>
  );
};

const defaultProps = {
  color: 'primary',
  hideDeleteIcon: false,
}

WishlistItem.defaultProps = defaultProps;

export default WishlistItem;
