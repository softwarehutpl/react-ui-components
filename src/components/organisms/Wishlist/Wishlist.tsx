import React from 'react';
import WishlistItem, { IWishlistCommon, IWishlistProduct } from './WishlistItem';

interface IWishlist extends IWishlistCommon {
  className?: string;
  products: IWishlistProduct[];
}

const Wishlist = ({ className, products, color, nameTextProps, priceTextProps, itemClassName, onDeleteIconClick, onItemClick, deleteIconColor, hideDeleteIcon }: IWishlist) => (
  <div className={className}>
    {products.map((product, index) => (
      <WishlistItem
        key={product.name + index}
        itemClassName={itemClassName}
        product={product}
        color={color}
        nameTextProps={nameTextProps}
        priceTextProps={priceTextProps}
        onDeleteIconClick={onDeleteIconClick}
        onItemClick={onItemClick}
        deleteIconColor={deleteIconColor}
        hideDeleteIcon={hideDeleteIcon}
      />
    ))}
  </div>
);

export default Wishlist;
