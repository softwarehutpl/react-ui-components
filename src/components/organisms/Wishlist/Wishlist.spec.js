import React from 'react';
import { create, act } from 'react-test-renderer';
import { render } from '@testing-library/react';
import 'jest-styled-components';
import { ThemeProvider } from 'styled-components';
import theme from '../../../common/theme';
import Wishlist from './Wishlist';
import WishlistItem, { StyledWishlistDeleteIcon } from './WishlistItem';
import wishlistProducts from '../../../common/mocks/wishlistProducts';
import setupIntersectionObserverMock from '../../../helpers/intersectionObserverMock';
import Text from '../../atoms/Text/Text';

describe('Wishlist component', () => {
  beforeEach(() => {
    setupIntersectionObserverMock();
  });

  it('should render correctly', () => {
    const tree = create(
      <ThemeProvider theme={theme}>
        <Wishlist products={wishlistProducts} />
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot()
  })

  it('should render wishlist item with proper product data', () => {
    const item = create(
      <ThemeProvider theme={theme}>
        <WishlistItem product={wishlistProducts[0]} />
      </ThemeProvider>
    );
    const instance = item.root;
    const info = instance.findByType(WishlistItem).findAllByType(Text);
    expect(info[0].props.children).toBe('Blue summer hat');
    expect(info[1].props.children).toBe('25.00 EUR');
  });

  it('should render wishlist item text in color based on color prop', () => {
    const { getByText } = render(
      <ThemeProvider theme={theme}>
        <Wishlist products={wishlistProducts} color="secondary" />
      </ThemeProvider>
    );
    const text = getByText('Blue summer hat');
    expect(text).toHaveStyle('color: #2196f3');
  });

  it('should render wishlist item name and price with custom props', () => {
    const { getByText } = render(
      <ThemeProvider theme={theme}>
        <Wishlist
          products={wishlistProducts}
          nameTextProps={{ fontColor: 'red', fontWeight: '800' }}
          priceTextProps={{ fontSize: 12, textAlign: 'center' }}
        />
      </ThemeProvider>
    );
    const name = getByText('Blue summer hat');
    const price = getByText('25.00 EUR');
    expect(name).toHaveStyle('color: red');
    expect(name).toHaveStyle('font-weight: 800');
    expect(price).toHaveStyle('font-size: 12px');
    expect(price).toHaveStyle('text-align: center');
  });

  it('should call onClick function', () => {
    const mockFn = jest.fn();
    const wishlist = create(
      <ThemeProvider theme={theme}>
        <Wishlist products={wishlistProducts} onItemClick={mockFn} />
      </ThemeProvider>
    )
    const instance = wishlist.root;
    const item = instance.findAllByType(WishlistItem)[0].findByType('div');
    act(() => {
      item.props.onClick();
    })
    expect(mockFn.mock.calls.length).toEqual(1);
  })

  it('should call onDeleteIconClick function', () => {
    const mockFn = jest.fn();
    const e = { stopPropagation: jest.fn() };
    const wishlist = create(
      <ThemeProvider theme={theme}>
        <Wishlist products={wishlistProducts} onDeleteIconClick={mockFn} />
      </ThemeProvider>
    )
    const instance = wishlist.root;
    const deleteIcon = instance.findAllByType(StyledWishlistDeleteIcon)[0];
    act(() => {
      deleteIcon.props.onClick(e);
    })
    expect(mockFn.mock.calls.length).toEqual(1);
  })
});
