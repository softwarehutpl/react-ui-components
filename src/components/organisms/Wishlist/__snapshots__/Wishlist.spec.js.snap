// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wishlist component should render correctly 1`] = `
.c1 {
  color: #000000;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  text-align: justify;
}

.c2 {
  color: #000000;
  font-size: 16px;
  line-height: 1.5;
  font-weight: normal;
  text-align: justify;
}

.c0 {
  object-fit: cover;
  width: 100%;
  height: 100%;
}

.c4 {
  display: inline-block;
  vertical-align: -.125em;
  overflow: hidden;
}

.c6 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c5 {
  color: #ffffff;
  background-color: #000000;
  width: 32px;
  padding: 7px;
  border-radius: 5px;
  -webkit-transition: -webkit-transform .2s;
  -webkit-transition: transform .2s;
  transition: transform .2s;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  cursor: pointer;
}

.c5:hover {
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c3 {
  height: 32px;
  width: 32px;
}

.c7 {
  color: #000000;
  width: 30px;
  position: absolute;
  top: 0;
  right: 0;
  cursor: pointer;
}

<div>
  <div
    className="undefined wishlistItem"
    onClick={[Function]}
  >
    <div
      className="wishlistItem__imageWrapper"
    >
      <img
        alt="Blue summer hat"
        className="c0"
        src="https://cdn.pixabay.com/photo/2015/10/13/03/39/fashion-985556_1280.jpg"
      />
    </div>
    <div
      className="wishlistItem__textWrapper"
    >
      <p
        className="c1"
      >
        Blue summer hat
      </p>
      <p
        className="c2"
      >
        25.00 EUR
      </p>
      <div
        className="c3 wishlistItem__cartIcon undefined"
        width={32}
      >
        <svg
          aria-hidden="true"
          className="c4 c5 cartIconContainer--nolabel"
          color="primary"
          fill="currentColor"
          focusable="false"
          onClick={[Function]}
          viewBox="0 0 576 512"
          width={32}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M504.717 320H211.572l6.545 32h268.418c15.401 0 26.816 14.301 23.403 29.319l-5.517 24.276C523.112 414.668 536 433.828 536 456c0 31.202-25.519 56.444-56.824 55.994-29.823-.429-54.35-24.631-55.155-54.447-.44-16.287 6.085-31.049 16.803-41.548H231.176C241.553 426.165 248 440.326 248 456c0 31.813-26.528 57.431-58.67 55.938-28.54-1.325-51.751-24.385-53.251-52.917-1.158-22.034 10.436-41.455 28.051-51.586L93.883 64H24C10.745 64 0 53.255 0 40V24C0 10.745 10.745 0 24 0h102.529c11.401 0 21.228 8.021 23.513 19.19L159.208 64H551.99c15.401 0 26.816 14.301 23.403 29.319l-47.273 208C525.637 312.246 515.923 320 504.717 320zM408 168h-48v-40c0-8.837-7.163-16-16-16h-16c-8.837 0-16 7.163-16 16v40h-48c-8.837 0-16 7.163-16 16v16c0 8.837 7.163 16 16 16h48v40c0 8.837 7.163 16 16 16h16c8.837 0 16-7.163 16-16v-40h48c8.837 0 16-7.163 16-16v-16c0-8.837-7.163-16-16-16z"
            fill="currentColor"
          />
        </svg>
      </div>
      <svg
        aria-hidden="true"
        className="c6 c7"
        color="primary"
        fill="currentColor"
        focusable="false"
        onClick={[Function]}
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M6 21h12V7H6v14zm2.46-9.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4h-3.5z"
        />
      </svg>
    </div>
  </div>
  <div
    className="undefined wishlistItem"
    onClick={[Function]}
  >
    <div
      className="wishlistItem__imageWrapper"
    >
      <img
        alt="Huge sombrero"
        className="c0"
        src="https://cdn.pixabay.com/photo/2015/12/08/01/04/sombrero-1082322_1280.jpg"
      />
    </div>
    <div
      className="wishlistItem__textWrapper"
    >
      <p
        className="c1"
      >
        Huge sombrero
      </p>
      <p
        className="c2"
      >
        30.99 EUR
      </p>
      <div
        className="c3 wishlistItem__cartIcon undefined"
        width={32}
      >
        <svg
          aria-hidden="true"
          className="c4 c5 cartIconContainer--nolabel"
          color="primary"
          fill="currentColor"
          focusable="false"
          onClick={[Function]}
          viewBox="0 0 576 512"
          width={32}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M504.717 320H211.572l6.545 32h268.418c15.401 0 26.816 14.301 23.403 29.319l-5.517 24.276C523.112 414.668 536 433.828 536 456c0 31.202-25.519 56.444-56.824 55.994-29.823-.429-54.35-24.631-55.155-54.447-.44-16.287 6.085-31.049 16.803-41.548H231.176C241.553 426.165 248 440.326 248 456c0 31.813-26.528 57.431-58.67 55.938-28.54-1.325-51.751-24.385-53.251-52.917-1.158-22.034 10.436-41.455 28.051-51.586L93.883 64H24C10.745 64 0 53.255 0 40V24C0 10.745 10.745 0 24 0h102.529c11.401 0 21.228 8.021 23.513 19.19L159.208 64H551.99c15.401 0 26.816 14.301 23.403 29.319l-47.273 208C525.637 312.246 515.923 320 504.717 320zM408 168h-48v-40c0-8.837-7.163-16-16-16h-16c-8.837 0-16 7.163-16 16v40h-48c-8.837 0-16 7.163-16 16v16c0 8.837 7.163 16 16 16h48v40c0 8.837 7.163 16 16 16h16c8.837 0 16-7.163 16-16v-40h48c8.837 0 16-7.163 16-16v-16c0-8.837-7.163-16-16-16z"
            fill="currentColor"
          />
        </svg>
      </div>
      <svg
        aria-hidden="true"
        className="c6 c7"
        color="primary"
        fill="currentColor"
        focusable="false"
        onClick={[Function]}
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M6 21h12V7H6v14zm2.46-9.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4h-3.5z"
        />
      </svg>
    </div>
  </div>
  <div
    className="undefined wishlistItem"
    onClick={[Function]}
  >
    <div
      className="wishlistItem__imageWrapper"
    >
      <img
        alt="Hunter hat"
        className="c0"
        src="https://cdn.pixabay.com/photo/2016/11/18/14/15/forest-1834831_1280.jpg"
      />
    </div>
    <div
      className="wishlistItem__textWrapper"
    >
      <p
        className="c1"
      >
        Hunter hat
      </p>
      <p
        className="c2"
      >
        23.50 EUR
      </p>
      <div
        className="c3 wishlistItem__cartIcon undefined"
        width={32}
      >
        <svg
          aria-hidden="true"
          className="c4 c5 cartIconContainer--nolabel"
          color="primary"
          fill="currentColor"
          focusable="false"
          onClick={[Function]}
          viewBox="0 0 576 512"
          width={32}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M504.717 320H211.572l6.545 32h268.418c15.401 0 26.816 14.301 23.403 29.319l-5.517 24.276C523.112 414.668 536 433.828 536 456c0 31.202-25.519 56.444-56.824 55.994-29.823-.429-54.35-24.631-55.155-54.447-.44-16.287 6.085-31.049 16.803-41.548H231.176C241.553 426.165 248 440.326 248 456c0 31.813-26.528 57.431-58.67 55.938-28.54-1.325-51.751-24.385-53.251-52.917-1.158-22.034 10.436-41.455 28.051-51.586L93.883 64H24C10.745 64 0 53.255 0 40V24C0 10.745 10.745 0 24 0h102.529c11.401 0 21.228 8.021 23.513 19.19L159.208 64H551.99c15.401 0 26.816 14.301 23.403 29.319l-47.273 208C525.637 312.246 515.923 320 504.717 320zM408 168h-48v-40c0-8.837-7.163-16-16-16h-16c-8.837 0-16 7.163-16 16v40h-48c-8.837 0-16 7.163-16 16v16c0 8.837 7.163 16 16 16h48v40c0 8.837 7.163 16 16 16h16c8.837 0 16-7.163 16-16v-40h48c8.837 0 16-7.163 16-16v-16c0-8.837-7.163-16-16-16z"
            fill="currentColor"
          />
        </svg>
      </div>
      <svg
        aria-hidden="true"
        className="c6 c7"
        color="primary"
        fill="currentColor"
        focusable="false"
        onClick={[Function]}
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M6 21h12V7H6v14zm2.46-9.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4h-3.5z"
        />
      </svg>
    </div>
  </div>
</div>
`;
